### PR TITLE
Convenience function to listen + accept + wait + exit

### PR DESCRIPTION
**Pull request by Ablu**
### Summary of the PR

We figured that all vhost-device daemons just ended up copying code from
each other. During an earlier attempt to solve that with a helper crate
for vhost-device [1], it was suggested to extend the vhost-user-backend
API directly. This way more people can benefit from this.

[1] https://github.com/rust-vmm/vhost-device/pull/362

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

### CODE REVIEW

**Comment by Ablu** [comment]
# Changelog
- Dropping DRAFT status
- Switched to AsRef<Path> for more flexible API
- Silence error messages which are "expected"
- Auto-delete socket files
- Posted vhost-device patch (https://github.com/rust-vmm/vhost-device/pull/389 -- commit: `tree-wide: use new serve convenience function`)

**Comment by Ablu** [comment]
# Changelog
- Added CHANGELOG.md entry

**Review by sboeuf — COMMENTED** [review]
Looks good, just one comment about the errors converted into success.
Factorization is always welcome :)

File: crates/vhost-user-backend/src/lib.rs
Comment (Ablu) [reply crates/vhost-user-backend/src/lib.rs:219]: > It might be appropriate to let the caller decide what to do with the result.

Hm... I intentionally added it here since every daemon in vhost-device checked for this and logged a not very helpful message... But I got no strong opinions here. If this only happens rarely, maybe we can also just return an error and tweak the error string a little bit? @vireshk do you see this message often?

Comment (sboeuf) [reply crates/vhost-user-backend/src/lib.rs:219]: IMO it doesn't matter if you see it often or not, it's a matter of letting the backend implementation decide if this is relevant or not.

Comment (sboeuf) [reply crates/vhost-user-backend/src/lib.rs:219]: I understand. But there's a difference between factorizing code and hiding something from the caller.
I'll let the other reviewers chime in on that topic, I'm okay with the current code if everybody else is okay with it.

Comment (Ablu) [reply crates/vhost-user-backend/src/lib.rs:219]: > hiding something from the caller

Sure! We should only do this if we are sure that these are usually innocent errors. Maybe there is some way to check whether the other end actually hung up after receiving a partial message...

Comment (Ablu) [reply crates/vhost-user-backend/src/lib.rs:219]: Any other opinions from other CODEOWNERS? :)

Comment (stefano-garzarella) [reply crates/vhost-user-backend/src/lib.rs:219]: That might be fine with me, but I would specify it better in the documentation that these errors are ignored as they are expected during frontend closure.

An alternative I can think of is to add a parameter to ignore them or not, but I'm not sure, in the end, if a backend wants to have complete control, it can use the functions that are already there.

Comment (Ablu) [reply crates/vhost-user-backend/src/lib.rs:219]: I will mention it in the docs.
I would like to avoid an additional parameter. My goal was to cover the 90% case with 10% of the code. If that does not work this way, I would go back to the drawing board.

**Comment by Ablu** [comment]
# Changelog
- rebased
- mention in the docs that disconnect related errors will be treated as success

**Review by stefano-garzarella — APPROVED** [review]
> # Changelog
> 
>     * rebased

Did you rebase on the latest main?
GH says: "This branch is out-of-date with the base branch"
Anyway I think we can update it before merge ;-)

> 
>     * mention in the docs that disconnect related errors will be treated as success

LGTM!

**Comment by Ablu** [comment]
> Did you rebase on the latest main?

Ah, I attempted the rebase before amending the commit -> the rebase refused to do anything with the dirty tree. Rebased now.

Thanks for the review!

File: crates/vhost-user-backend/src/lib.rs
Comment (Ablu) [reply crates/vhost-user-backend/src/lib.rs:222]: @sboeuf: Does the update with explicitly documenting this works for you?

Comment (sboeuf) [reply crates/vhost-user-backend/src/lib.rs:222]: Yes looks good

Comment (Ablu) [reply crates/vhost-user-backend/src/lib.rs:222]: Are you happy to review the PR then? :innocent:

**Review by sboeuf — APPROVED** [review]

**Comment by Ablu** [comment]
I had to rebase on top 282a10f8e12883ecf8e16c49b41223ad726f2fd2. Sorry, need to re-request reviews.

**Comment by stefano-garzarella** [comment]
@Ablu sorry, another rebase required (I did it). I can merge it after CI completed!

**Comment by Ablu** [comment]
> @Ablu sorry, another rebase required (I did it). I can merge it after CI completed!

No worries, I think "clean"/automatic rebases also do not reset the reviews, but I had to resolve a conflict due to the renaming.

**Review by stefano-garzarella — APPROVED** [review]

**Pull request merged**
